src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs

```diff
@@ -1713,12 +1713,21 @@ public static TypeSyntax ParseTypeName(string text, int offset, bool consumeFull
         /// </summary>
         public static TypeSyntax ParseTypeName(string text, int offset = 0, ParseOptions? options = null, bool consumeFullText = true)
         {
-            using (var lexer = MakeLexer(text, offset, (CSharpParseOptions?)options))
+            var cSharpOptions = (CSharpParseOptions?)options;
+            using (var lexer = MakeLexer(text, offset, cSharpOptions))
             using (var parser = MakeParser(lexer))
             {
                 var node = parser.ParseTypeName();
                 if (consumeFullText) node = parser.ConsumeUnexpectedTokens(node);
-                return (TypeSyntax)node.CreateRed();
+                var red = (TypeSyntax)node.CreateRed();
+                if (cSharpOptions is not null)
```

Comment:
Feels like we should do this unconditionally, and use `lexer.Options`, to ensure that this always stays in sync. We could also consider factoring this part out into a helper for all the `ParseX` helpers:

```cs
private TSyntax CreateRed<TSyntax>(SyntaxNode node, CSharpParseOptions options)
{
    var red = (TSyntax)(object)node;
#pragma warning disable RS0030 // Do not use banned APIs (CreateWithoutClone is intended to be used from this call site)
    red._syntaxTree = CSharpSyntaxTree.CreateWithoutClone(red, options);
#pragma warning restore
    return red;
}
```

Called as:

```cs
CreateRed<TypeSyntax>(node, lexer.Options);
```

Up to you if you want to refactor that far though.